sacred/ingredient.py

```diff
@@ -351,11 +355,18 @@ def traverse_ingredients(self):
             If a circular structure among ingredients was detected.
         """
         if self._is_traversing:
-            raise CircularDependencyError()
+            raise CircularDependencyError(ingredients=[self])
         else:
             self._is_traversing = True
         yield self, 0
-        for ingredient in self.ingredients:
-            for ingred, depth in ingredient.traverse_ingredients():
-                yield ingred, depth + 1
+        try:
+            for ingredient in self.ingredients:
+                for ingred, depth in ingredient.traverse_ingredients():
+                    yield ingred, depth + 1
+        except CircularDependencyError as e:
```

Comment:
You use this pattern a lot. Basically:
``` python
try:
   #  stuff
except SomeCustomError as e:
  # special handling (i.e. adding information) of that error
```
I think it might be nicer to provide a context manager to provide that special handling without cluttering the code too much. It could even be part of the exception itself. Something like:
``` python
with CircularDependencyError.track(self):  # not sure about the name
    # stuff
```
Shouldn't be hard to implement and would improve readability IMHO.

Reply:
I already thought about doing this. This should be easy to implement and improve readability a lot.